File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.py
# Last updated: 2/28/2026, 4:27:17 PM
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:


    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        maxDiameter = [0]

        self.traverse(root,maxDiameter)

        return maxDiameter[0]

    def traverse(self, root: Optional[TreeNode], maxDiameter: list) -> int:

        if not root: return 0

        l = self.traverse(root.left,maxDiameter)
        r = self.traverse(root.right,maxDiameter)

        maxDiameter[0] = max(maxDiameter[0], l + r)

        return max(l,r) + 1




        

File: 543-DiameterOfBinaryTree/543-DiameterOfBinaryTree.java
// Last updated: 2/28/2026, 4:27:19 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int diameterOfBinaryTree(TreeNode root) {
        
        int max[] = new int[1];

        traverse(root, max);

        return max[0];

    }


    public int traverse(TreeNode root, int max[]){

        if(root == null) return 0;

        int l = traverse(root.left, max);
        int r = traverse(root.right,max);

        max[0] = Math.max(l + r, max[0]);

        return Math.max(l , r) + 1;

    }
}